### PR TITLE
Ibalazevic assignment2

### PR DESCRIPTION
**Pull request by ibalazevic**

### CODE REVIEW

File: ibalazevic/assignment2/q1.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        node.right_child = Node(data)
+
+
+    def print_ancestors(self, node, key):
+        """
+        A method for printing all ancestors of a given key.
+            - node - Node, the root node of the tree.
+            - key - int, the node for which we are printing
+                    the ancestors.
+            Returns: prints the ancestors. If a key is not
```
[the remark this answers is not in the document]
Comment (ibalazevic) [reply ibalazevic/assignment2/q1.py]: That's how I started off, but I thought it doesn't make sense to store a list into memory if we don't have to, just for the purpose of testing. What do you think?

File: ibalazevic/assignment2/q1.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+                     present in the tree, returns True,
+                     False otherwise.
+        """
+        if not node:
+            return False
+        if node.data == key:
+            return True
+
+        if self.print_ancestors(node.left_child, key) or\
```
[the remark this answers is not in the document]
Comment (ibalazevic) [reply ibalazevic/assignment2/q1.py]: Done

File: ibalazevic/assignment2/q1.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    def insert_right(self, node, data):
+        """
+        A method for inserting a right child of a node.
+            - node - Node, the current node.
+            - data - int, the data to insert to the right.
+        """
+        node.right_child = Node(data)
+
+
+    def print_ancestors(self, node, key):
```
[the remark this answers is not in the document]
Comment (ibalazevic) [reply ibalazevic/assignment2/q1.py]: Done

File: ibalazevic/assignment2/q1.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
+import unittest
+
+class Node:
+    """
+    A class for a node of a binary tree.
+        - data - int, the content of a node.
```
[the remark this answers is not in the document]
Comment (ibalazevic) [reply ibalazevic/assignment2/q1.py]: Done

File: ibalazevic/assignment2/q2.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+
+class BinaryTree:
+    """
+    A class representing a binary tree.
+        - root_data - int, content of the tree root.
+    """
+    def __init__(self, root_data):
+        self.root = Node(root_data)
+        self.ancestor = self.root.data
```
[the remark this answers is not in the document]
Comment (ibalazevic) [reply ibalazevic/assignment2/q2.py]: Oops, no, I forgot to delete them. Done.

File: ibalazevic/assignment2/q2.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
+import unittest
```
[the remark this answers is not in the document]
Comment (ibalazevic) [reply ibalazevic/assignment2/q2.py]: Ack

File: ibalazevic/assignment2/q2.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+    def insert_right(self, node, data):
+        """
+        A method for inserting a right child of a node.
+            - node - Node, the current node.
+            - data - int, the data to insert to the right.
+        """
+        node.right_child = Node(data)    
+
+    def lowest_common_ancestor(self, node, key1, key2):
```
[the remark this answers is not in the document]
Comment (ibalazevic) [reply ibalazevic/assignment2/q2.py]: Done

File: ibalazevic/assignment2/q1.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    def __init__(self, root_data):
+        self.root = Node(root_data)
+
+    def insert_left(self, node, data):
+        """
+        A method for inserting a left child of a node.
+            - node - Node, the current node.
+            - data - int, the data to insert to the left.
+        """
+        node.left_child = Node(data)
```
[the remark this answers is not in the document]
Comment (ibalazevic) [reply ibalazevic/assignment2/q1.py]: Added a comment saying that I assume there are no duplicates. Done

File: ibalazevic/assignment2/q1.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ibalazevic) [reply ibalazevic/assignment2/q1.py]: Done

File: ibalazevic/assignment2/q1.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        self.tree.insert_right(self.tree.root.left_child, 5)
+        self.tree.insert_right(self.tree.root.right_child, 8)
+        self.tree.insert_left(self.tree.root.left_child.left_child, 1)
+        self.tree.insert_right(self.tree.root.left_child.left_child, 6)
+   
+
+    def test_root(self):
+        self.assertEqual(self.tree.print_ancestors(self.tree.root, 7), True)
+
+    def test_middle(self):
```
[the remark this answers is not in the document]
Comment (ibalazevic) [reply ibalazevic/assignment2/q1.py]: Done

File: ibalazevic/assignment2/q2.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        """
+        A method for finding the lowest common ancestor of
```
[the remark this answers is not in the document]
Comment (ibalazevic) [reply ibalazevic/assignment2/q2.py]: Same as above - added a comment saying that I assume there are no duplicates. Done

File: ibalazevic/assignment2/q2.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        two nodes in a binary tree.
+            - node - Node, the root node of the tree.
+            - key1 - int, value of first of the nodes.
+            - key2 - int, value of second of the nodes.
+            Returns: int, value of the lowest common
+                     ancestor node.
+        """
+        if not node:
+            return None
```
[the remark this answers is not in the document]
Comment (ibalazevic) [reply ibalazevic/assignment2/q2.py]: Done

File: ibalazevic/assignment2/q2.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+        if node.data == key1 or node.data == key2:
+            return node.data
+
+        left_subtree = self.lowest_common_ancestor(node.left_child, key1, key2)
+        right_subtree = self.lowest_common_ancestor(node.right_child, key1, key2)
+
+        if left_subtree and right_subtree:
```
[the remark this answers is not in the document]
Comment (ibalazevic) [reply ibalazevic/assignment2/q2.py]: Done

File: ibalazevic/assignment2/q2.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+        self.tree = BinaryTree(7)
+        self.tree.insert_left(self.tree.root, 3)
+        self.tree.insert_right(self.tree.root, 4)
+        self.tree.insert_left(self.tree.root.left_child, 2)
+        self.tree.insert_right(self.tree.root.left_child, 5)
+        self.tree.insert_right(self.tree.root.right_child, 8)
+        self.tree.insert_left(self.tree.root.left_child.left_child, 1)
+        self.tree.insert_right(self.tree.root.left_child.left_child, 6)
+   
+
```
[the remark this answers is not in the document]
Comment (ibalazevic) [reply ibalazevic/assignment2/q2.py]: Done

File: ibalazevic/assignment2/q1.py [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+
+class BinaryTree:
+    """
+    A class representing a binary tree.
+        - root_data - int, content of the tree root.
+    """
+    def __init__(self, root_data):
+        self.root = Node(root_data)
+
+    def insert_left(self, node, data):
```
[the remark this answers is not in the document]
Comment (ibalazevic) [reply ibalazevic/assignment2/q1.py]: Ack

File: ibalazevic/assignment2/q1.py [hunk 15]
```diff
@@ -26,39 +28,57 @@ def insert_left(self, node, data):
...
+                    the ancestors.
+            Returns: list of ancestors.
+           
+        """
+        ancestors = self.find_ancestors(self.root, key, [])
+        if ancestors is not None:
+            print ", ".join(map(str, ancestors))
+            return ancestors
+        else:
+            raise ValueError("Key not present in the binary tree.")
```
Comment (ferreron) [inline ibalazevic/assignment2/q1.py]: Probably "KeyError" makes more sense in this case. From the Python documentation:
KeyError: Raised when a mapping (dictionary) key is not found in the set of existing keys.
ValueError: Raised when a built-in operation or function receives an argument that has the right type but an inappropriate value, and the situation is not described by a more precise exception such as IndexError.

File: ibalazevic/assignment2/q1.py [hunk 16]
```diff
@@ -26,39 +28,57 @@ def insert_left(self, node, data):
...
+
+    def find_ancestors(self, node, key, ancestors):
```
Comment (ferreron) [inline ibalazevic/assignment2/q1.py]: If this method is meant to be private, I would name it `_find_ancestors`, with the leading underscore. See: https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles

File: ibalazevic/assignment2/q2.py [hunk 17]
```diff
@@ -15,41 +15,77 @@ def __init__(self, data):
...
+            - key1 - value of first of the nodes.
+            - key2 - value of second of the nodes.
+            Returns: value of the lowest common
+                     ancestor node, raises ValueError if
+                     either one of the keys is not present
+                     in the tree.
+        """
+        if not (self.find_node(self.root, key1) and
+                self.find_node(self.root, key2)):
+            raise ValueError("Key not present in the binary tree.")
```
Comment (ferreron) [inline ibalazevic/assignment2/q2.py]: Same as above, KeyError seems more appropriate.

File: ibalazevic/assignment2/q1.py (hunk 16 above)
Comment (ibalazevic) [reply ibalazevic/assignment2/q1.py]: Done

**Pull request merged**